internal/catalog/internal/controllers/workloadhealth/controller.go

```diff
@@ -25,95 +27,49 @@ var (
 )
 
-// The NodeMapper interface is used to provide an implementation around being able to
-// map a watch event for a Node resource and translate it to reconciliation requests
-// for all Workloads assigned to that node.
-type NodeMapper interface {
-	// MapNodeToWorkloads will take a Node resource and return controller requests
-	// for all Workloads associated with the Node.
-	MapNodeToWorkloads(ctx context.Context, rt controller.Runtime, res *pbresource.Resource) ([]controller.Request, error)
-
-	// TrackWorkload instructs the NodeMapper to associate the given workload
-	// ID with the given node ID.
-	TrackWorkload(workloadID *pbresource.ID, nodeID *pbresource.ID)
-
-	// UntrackWorkload instructs the Nodemapper to forget about any
-	// association it was tracking for this workload.
-	UntrackWorkload(workloadID *pbresource.ID)
-
-	// NodeIDFromWorkload is used to generate the resource ID for the Node referenced
-	// within the NodeName field of the Workload.
-	NodeIDFromWorkload(workload *pbresource.Resource, workloadData *pbcatalog.Workload) *pbresource.ID
-}
-
-func WorkloadHealthController(nodeMap NodeMapper) *controller.Controller {
-	if nodeMap == nil {
-		panic("No NodeMapper was provided to the WorkloadHealthController constructor")
-	}
+func WorkloadHealthController() *controller.Controller {
+	nodeIndexer := indexers.DecodedSingleIndexer[*pbcatalog.Workload](
+		"node",
+		index.ReferenceOrIDFromArgs,
+		nodeIndexer,
+	)
 
-	return controller.NewController(StatusKey, pbcatalog.WorkloadType).
+	return controller.NewController(StatusKey, pbcatalog.WorkloadType, nodeIndexer).
 		WithWatch(pbcatalog.HealthStatusType, dependency.MapOwnerFiltered(pbcatalog.WorkloadType)).
-		WithWatch(pbcatalog.NodeType, nodeMap.MapNodeToWorkloads).
-		WithReconciler(&workloadHealthReconciler{nodeMap: nodeMap})
+		WithWatch(pbcatalog.NodeType, dependency.CacheListMapper(pbcatalog.WorkloadType, nodeIndexer.Name())).
+		WithReconciler(&workloadHealthReconciler{})
 }
 
-type workloadHealthReconciler struct {
-	nodeMap NodeMapper
-}
+type workloadHealthReconciler struct{}
 
 func (r *workloadHealthReconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
 	// The runtime is passed by value so replacing it here for the remainder of this
 	// reconciliation request processing will not affect future invocations.
-	rt.Logger = rt.Logger.With("resource-id", req.ID, "controller", StatusKey)
+	rt.Logger = rt.Logger.With("resource-id", req.ID)
 
 	rt.Logger.Trace("reconciling workload health")
 
 	// read the workload
-	rsp, err := rt.Client.Read(ctx, &pbresource.ReadRequest{Id: req.ID})
-	switch {
-	case status.Code(err) == codes.NotFound:
-		rt.Logger.Trace("workload has been deleted")
-		r.nodeMap.UntrackWorkload(req.ID)
-		return nil
-	case err != nil:
+	workload, err := resource.GetDecodedResource[*pbcatalog.Workload](ctx, rt.Client, req.ID)
```

Comment:
`GetDecodedResource` returns `nil` instead of returning an error when the resource is not found. The updated code does not seem to handle `workload == nil` in the same way.

Reply:
You are right and the integration tests were catching this as nil pointer panics. It has been fixed.